### PR TITLE
Allow T op= &T for built-in numeric types T v2

### PR DESCRIPTION
**Pull request by Eh2406**
Manually rebase of @Migi https://github.com/rust-lang/rust/pull/41336

### CODE REVIEW

**Comment by kennytm** [comment]
The [`for-loop-unconstrained-element-type-i32-fallback.rs` test](https://github.com/rust-lang/rust/blob/master/src/test/run-pass/for-loop-unconstrained-element-type-i32-fallback.rs) failed (probably not essential for crater).

```
[00:50:26] ---- [run-pass] run-pass/for-loop-unconstrained-element-type-i32-fallback.rs stdout ----
[00:50:26] 	
[00:50:26] error: compilation failed!
[00:50:26] status: exit code: 101
[00:50:26] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-pass/for-loop-unconstrained-element-type-i32-fallback.rs" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass" "--target=x86_64-unknown-linux-gnu" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/for-loop-unconstrained-element-type-i32-fallback.stage2-x86_64-unknown-linux-gnu" "-Crpath" "-O" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/for-loop-unconstrained-element-type-i32-fallback.stage2-x86_64-unknown-linux-gnu.run-pass.libaux"
[00:50:26] stdout:
[00:50:26] ------------------------------------------
[00:50:26] 
[00:50:26] ------------------------------------------
[00:50:26] stderr:
[00:50:26] ------------------------------------------
[00:50:26] error[E0282]: type annotations needed
[00:50:26]   --> /checkout/src/test/run-pass/for-loop-unconstrained-element-type-i32-fallback.rs:17:5
[00:50:26]    |
[00:50:26] 17 |        for i in Vec::new() {
[00:50:26]    |   _____^
[00:50:26]    |  |_____|
[00:50:26]    | ||
[00:50:26] 18 | ||         sum += i;
[00:50:26] 19 | ||     }
[00:50:26]    | ||     ^
[00:50:26]    | ||_____|
[00:50:26]    | |______cannot infer type for `_`
[00:50:26]    |        consider giving `__next` a type
[00:50:26] 
[00:50:26] error: aborting due to previous error
```

**Comment by Eh2406** [comment]
I'm not seeing how to fix the test. It is a very fragile situation in which this test passes. I.E. this [`sum = sum + i;`](http://play.integer32.com/?gist=0f3d5bc7a8b3a191f9bb01bdb077e173) does not work on stable.

**Comment by kennytm** [comment]
The test was introduced in https://github.com/rust-lang/rust/pull/42634#issuecomment-308560533 (cc @nikomatsakis). Maybe you could just disable the test (add a comment `// ignore-test`) to make the CI green for now.

**Comment by Eh2406** [comment]
Thanks for suggestion. Let's see what ci thinks.

**Comment by aturon** [comment]
@rust-lang/infra cargobomb requested!

**Comment by Mark-Simulacrum** [comment]
Started preparing cargobomb run.

File: src/test/run-pass/op-assign-builtins-by-ref.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        assert_eq!(x, 0b101000u64);
+    }
+
+    // test ShrAssign
+    {
+        let mut x = 0b1010u64;
+        x >>= &2i16;
+        assert_eq!(x, 0b10u64);
+    }
+}
```
Comment (bjorn3) [inline src/test/run-pass/op-assign-builtins-by-ref.rs]: Missing trailing newline

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: src/test/run-pass/op-assign-builtins-by-ref.rs (hunk 1 above)
Comment (Eh2406) [reply src/test/run-pass/op-assign-builtins-by-ref.rs]: Should I fix now or wait for cargobomb?

Comment (kennytm) [reply src/test/run-pass/op-assign-builtins-by-ref.rs]: @Eh2406 Cargobomb uses the compiler already built from `@bors try`. You could fix it any time you like.

Comment (Eh2406) [reply src/test/run-pass/op-assign-builtins-by-ref.rs]: Done, thanks!

**Comment by Mark-Simulacrum** [comment]
Cargobomb run complete: http://cargobomb-reports.s3.amazonaws.com/pr-44287/index.html.

**Comment by Eh2406** [comment]
Thanks for running this! I am new to cargobomb, what is the best way to proceed? How can I find the original code, so I can start on Pr's?

Any iday how this Pr can lead to a `test-fail` regression?

**Comment by kennytm** [comment]
@Eh2406:

* **regressed** = Previously success, now failed. Focus on this.
* **fixed** = Previously failed, now success. Likely badly written tests in this case.
* **unknown** = Unknown. 
* **build-fail/test-fail** = Failed in both toolchains. That means the code itself is buggy
* **test-pass** = success in both toolchains. That's good.

Usually **regressed** is the only section you need to pay attention to.

Since this change affects type inference, the "test-fail" results likely means the crate's tests are flaky, and can be safely ignored. The list of crates with flaky tests can be found in https://github.com/rust-lang-nursery/cargobomb/blob/master/blacklist.md, rust-lang-nursery/cargobomb#126 and rust-lang-nursery/cargobomb#127. For instance `ai-0.1.0` is known to be flaky (#43963). 

Sometimes even "build-fail" tests are spurious, e.g. `rust-lang.cargo.98b6f6…` failed due to rust-lang-nursery/cargobomb#125, not real regression.

However, the error in `generator-0.5.0` is caused by failure to build the example, but wrongly attributed as a "test-fail". In the end you need to read every log to be sure.

---

There are 12 root regressions, much more than [the previous try](https://github.com/rust-lang/rust/pull/41336#issuecomment-304766078).

| Crate | Code |
|---|---|
| average | `self.n[i] += d.approx().unwrap();` |
| cov | `exit_count -= self.graph` |
| etcommon-trie | `ret[end] \|= vec[i].into();` |
| finch | `total_gc += kmer.kmer.iter().map(\|b\| {` |
| gene-seq-intersections | `tmp_count += x.into();` |
| generator | https://github.com/Xudong-Huang/generator-rs/blob/164e28dc/examples/send.rs#L5-L15 |
| libcantal | `offset += size;` |
| minions | `sum += i;` |
| netlink-rs | `self.flags \|= Flags::Multi.into();` |
| treeflection_derive | `self.qux += value.parse().unwrap();` |
| whiteread | [`s += x`](https://github.com/krdln/whiteread/blob/1c2b2b3939/src/reader.rs#L582) |
| samsartor/pbr-demo@8cb72ed951165f4f9f73096d2e275f502d3a6635 | `rgb[3] *= v.parse().map_err(\|_\| "…")?` |

**Comment by kennytm** [comment]
I'd also like to note that rust-lang/rfcs#2147 may make this PR unnecessary.

**Comment by Migi** [comment]
@kennytm: Thanks a lot for the nice summary of the regressions. Does cargobomb test more crates than crater did? Because I find it surprising that there would be more people writing this kind of pattern in the last 3 months than in the 2 years before that.

But to answer your remark that "rust-lang/rfcs#2147 may make this PR unnecessary": I don't think that's true. In fact, I think actually the opposite may be true: rust-lang/rfcs#2147 may make the missing impls even more sorely missed. Let me explain:

In non-generic code, this PR and rust-lang/rfcs#2147 are two completely separate things. This PR is pretty useless in non-generic code, while rust-lang/rfcs#2147 makes it more ergonomic to do numerical computations on types that are not `Copy` and expensive to clone (such as `BigNum`). With those types, you want to use as many references as possible (to avoid cloning/moving), and rust-lang/rfcs#2147 makes you not have to write `&`s for all those references.

In generic code, rust-lang/rfcs#2147 is not so different from its non-generic version. Similar to how it makes it easier to use `BigNum`s, it makes it easier to use generic types that are not constrained to be `Copy`. For example, if I want to write a function that works both with `i64`s and with `BigNum`s, then rust-lang/rfcs#2147 would help me. But without this PR, I cannot use any op-assign operators (`+=`, `-=`, etc.) in that generic function, because `i64` doesn't satisfy the trait bound `T : AddAssign<&'a T>`. I can only use the normal binary operators `+`, `-`, `*`, etc.

So in conclusion, if rust-lang/rfcs#2147 is accepted, I don't think fewer people would want this PR to be accepted. I think more people would start doing generic numerical computations, and as a result, *more* people would want this PR to be accepted.

**Comment by Eh2406** [comment]
@Migi Thank you for the detailed analysis! And yes, cargobomb tests more than crater did. Crater just tried to build all the crates, Cargobomb tries to run all the tests. So the the regressions in test code, like treeflection_derive, would not have been caught by Crater.

@kennytm Thank you for the explanation and for doing the triage. I will work on making PR's to fix them later today/as soon as I have time.

How can I get rustup to use the try artifact, so I can double check my fixes?

**Comment by kennytm** [comment]
@Eh2406 Not sure how to setup rustup for this, but for could download the binaries from

https://rust-lang-ci.s3.amazonaws.com/rustc-builds-try/3c27d9e8363646c0da43e8adc7cd5cd4c84aa9d1/rust-nightly-x86_64-unknown-linux-gnu.tar.xz

**Comment by Mark-Simulacrum** [comment]
https://github.com/Mark-Simulacrum/bisect-rust will download and unarchive a try build, you just need to run `cargo run --bin install-sysroot --commit 3c27d9e8363646c0da43e8adc7cd5cd4c84aa9d1` with the try commit and everything should "just work" from there, I think. If not, please file an issue!

**Comment by Eh2406** [comment]
So PR's have been made; thanks again to @kennytm for the grate list,

I added a commit that re-adds and hopefully fixes the for-loop-unconstrained-element-type-i32-fallback.rs test. So if CI goes green, what is the next steps? :-)

**Comment by aturon** [comment]
@Eh2406 Next step is: let's **merge** this thing! Thanks so much for the work you put into it.

@bors r+

**Comment by scottmcm** [comment]
:tada: Hooray!  This is literally the [first](https://users.rust-lang.org/t/why-isnt-i32-addassign-i32/8218) major problem I ran into with Rust :tada:

**Comment by scottmcm** [comment]
```
   Compiling xsv v0.10.3 (file:///C:/projects/rust/build/ct/xsv)
error[E0283]: type annotations required: cannot resolve `f64: std::ops::AddAssign<_>`
   --> src\cmd\stats.rs:470:32
    |
470 |                         *float += from_bytes(sample).unwrap();
    |                                ^^
```

Seems to have been [updated in xsv master](https://github.com/BurntSushi/xsv/blob/master/src/cmd/stats.rs#L482); maybe just needs a dep version bump?

**Comment by Eh2406** [comment]
What is involved in a dep version bump? Is it just changing https://github.com/rust-lang/rust/blob/master/src/tools/cargotest/main.rs#L52 to the hash of head?

File: src/libcore/internal_macros.rs [hunk 2]
```diff
@@ -68,3 +68,22 @@ macro_rules! forward_ref_binop {
         }
     }
 }
+
+// implements "T op= &U", based on "T op= U"
+// where U is expected to be `Copy`able
+macro_rules! forward_ref_op_assign {
+    (impl $imp:ident, $method:ident for $t:ty, $u:ty) => {
+        forward_ref_op_assign!(impl $imp, $method for $t, $u,
+                #[stable(feature = "op_assign_builtins_by_ref", since = "1.18.0")]);
```
Comment (ollie27) [inline src/libcore/internal_macros.rs]: This will need to be `"1.22.0"`.

Comment (Eh2406) [reply src/libcore/internal_macros.rs]: Changed, thanks!

**Comment by Eh2406** [comment]
Not having heard back, I updated xsv by changing the sha to the head on github. I think the next step if for someone to r+ again or for someone to tell me how to correctly fix the xsv issue.

**Comment by aturon** [comment]
Thanks!

@bors: r+

**Comment by shepmaster** [comment]
@Bors r-

**Comment by Mark-Simulacrum** [comment]
Merged in https://github.com/rust-lang/rust/pull/44936.

**Comment by SimonSapin** [comment]
This broke Servo:

```
error[E0283]: type annotations required: cannot resolve `usize: core::ops::AddAssign<_>`
   --> /home/travis/build/servo/servo-with-rust-nightly/servo/components/script/textinput.rs:239:31
    |
239 |             |len, slice| *len += slice.chars().map(char::len_utf16).sum())
    |                               ^^
error: aborting due to previous error
```

(Using `.sum::<usize>()` fixes it, though it’s not pretty: https://github.com/servo/servo/pull/18693)

**Comment by shepmaster** [comment]
[Ditto fuzzy-pickles](https://github.com/shepmaster/fuzzy-pickles/issues/92):

```
error[E0283]: type annotations required: cannot resolve `usize: std::ops::AddAssign<_>`
   --> src/tokenizer.rs:544:17
    |
544 |             idx += ci.take_while(|&c| UnicodeXID::is_xid_continue(c)).map(|c| c.len_utf8()).sum();
    |                 ^^
```

**Comment by pnkfelix** [comment]
Given #45480, should this be tagged as `[breaking-change]` ?

**Comment by Migi** [comment]
I would be in favor of tagging this as `[breaking-change]`, since it is breaking some amount of real-world code.

However, it should be noted that type inference is specifically excluded from the backwards compatibility guarantees of Rust. See for example [in the official FAQ](https://www.rust-lang.org/en-US/faq.html), under "How does Rust language versioning work?":

> Rust’s language versioning follows SemVer, with backwards incompatible changes of stable APIs only allowed in minor versions if those changes fix compiler bugs, patch safety holes, or change dispatch **or type inference to require additional annotation.**

That FAQ further says "For additional details, read [the Rust blog post “Stability as a Deliverable.”](https://blog.rust-lang.org/2014/10/30/Stability.html), which also says the same, under "What are the stability caveats?":

> We reserve the right to fix compiler bugs, patch safety holes, and **change type inference in ways that may occasionally require new type annotations.**

So essentially, code relying on type inference is occasionally going to need some additional annotations. The reason is that type inference tries to do as much as it can, relying both on what impls exist and also on what impls *don't* exist. This is very useful, but has the downside that adding an impl may occasionally break working code, which is what happened here.

**Comment by SimonSapin** [comment]
Another important aspect of inference breakage is that it is (always?) possible to write code with more type annotations that works both in compiler with and without the change.
